src/codegen.cpp

```diff
@@ -6569,93 +6582,99 @@ static void emit_stmtpos(jl_codectx_t &ctx, jl_value_t *expr, int ssaval_result)
-    //    it = ctx.emission_context.compiled_functions.insert(std::make_pair(ci, std::make_pair(std::move(closure_m), std::move(closure_decls)))).first;
-    //}
-
-    //auto &closure_m = it->second.first;
-    //auto &closure_decls = it->second.second;
-
-    //assert(closure_decls.functionObject != "jl_fptr_sparam");
-    //bool isspecsig = closure_decls.functionObject != "jl_fptr_args";
-
-    //Function *F = NULL;
-    //std::string fname = isspecsig ?
-    //    closure_decls.functionObject :
-    //    closure_decls.specFunctionObject;
-    //if (GlobalValue *V = jl_Module->getNamedValue(fname)) {
-    //    F = cast<Function>(V);
-    //}
-    //else {
-    //    F = Function::Create(get_func_sig(ctx.builder.getContext()),
-    //                         Function::ExternalLinkage,
-    //                         fname, jl_Module);
-    //    jl_init_function(F, ctx.emission_context.TargetTriple);
-    //    jl_name_jlfunc_args(ctx.emission_context, F);
-    //    F->setAttributes(AttributeList::get(ctx.builder.getContext(), {get_func_attrs(ctx.builder.getContext()), F->getAttributes()}));
-    //}
-    //Function *specF = NULL;
-    //if (!isspecsig) {
-    //    specF = F;
-    //}
-    //else {
-    //    //emission context holds context lock so can get module
-    //    specF = closure_m.getModuleUnlocked()->getFunction(closure_decls.specFunctionObject);
-    //    if (specF) {
-    //        jl_returninfo_t returninfo = get_specsig_function(ctx, jl_Module, NULL,
-    //            closure_decls.specFunctionObject, sigtype, rettype, true, JL_FEAT_TEST(ctx,gcstack_arg));
-    //        specF = cast<Function>(returninfo.decl.getCallee());
-    //    }
-    //}
-    //JL_GC_POP();
-    //return std::make_pair(F, specF);
+    jl_svec_t *sig_args = NULL;
+    jl_value_t *sigtype = NULL;
+    JL_GC_PUSH2(&sig_args, &sigtype);
+
+    size_t nsig = 1 + jl_svec_len(argt_typ->parameters);
+    sig_args = jl_alloc_svec_uninit(nsig);
+    jl_svecset(sig_args, 0, env_t);
+    for (size_t i = 0; i < jl_svec_len(argt_typ->parameters); ++i) {
+        jl_svecset(sig_args, 1+i, jl_svecref(argt_typ->parameters, i));
+    }
+    sigtype = jl_apply_tuple_type_v(jl_svec_data(sig_args), nsig);
+
+    jl_method_instance_t *mi;
+    jl_code_instance_t *ci;
+
+    if (closure_method->source) {
+        mi = jl_specializations_get_linfo(closure_method, sigtype, jl_emptysvec);
+        ci = (jl_code_instance_t*)jl_rettype_inferred_addr(mi, ctx.min_world, ctx.max_world);
+    }
+    else {
+        mi = (jl_method_instance_t*)jl_atomic_load_relaxed(&closure_method->specializations);
+        assert(jl_is_method_instance(mi));
+        ci = jl_atomic_load_relaxed(&mi->cache);
+    }
+    if (ci == NULL || (jl_value_t*)ci == jl_nothing || ci->rettype != rettype || !jl_egal(sigtype, mi->specTypes)) { // TODO: correctly handle the ABI conversion if rettype != ci->rettype
+        JL_GC_POP();
+        return std::make_pair((Function*)NULL, (Function*)NULL);
```

Comment:
Out of curiosity, what does it mean when codegen fails to produce a result like this?

Reply:
codegen is always allowed to decline to generate optimized code, and the runtime must find some other way to execute it (usually in the interpreter)